File: src/lib/blocklist.ts
/**
 * Inbound (create-time) phishing/malware blocklist (ARCHITECTURE.md §4.5,
 * FR-36, AC-44). Offline newline-delimited host file loaded once into an
 * in-memory set. This is the INBOUND boundary — entirely separate from the
 * outbound SSRF guard (lib/ssrf.ts). Pure matching logic is exported and tested
 * independently of file loading.
 */
import { readFileSync, existsSync } from 'node:fs'
import { join } from 'node:path'

let cachedSet: Set<string> | null = null

/** Parse blocklist file text into a normalized host set. Pure. */
export function parseBlocklist(text: string): Set<string> {
  const set = new Set<string>()
  for (const line of text.split('\n')) {
    const trimmed = line.trim()
    if (trimmed === '' || trimmed.startsWith('#')) continue
    set.add(normalizeHost(trimmed))
  }
  return set
}

function normalizeHost(host: string): string {
  return host.toLowerCase().replace(/^www\./, '').replace(/\.$/, '')
}

function loadSet(): Set<string> {
  if (cachedSet) return cachedSet
  const path = join(process.cwd(), 'data', 'blocklist.txt')
  if (existsSync(path)) {
    try {
      cachedSet = parseBlocklist(readFileSync(path, 'utf8'))
      return cachedSet
    } catch {
      /* fall through to empty set */
    }
  }
  cachedSet = new Set()
  return cachedSet
}

/**
 * Pure: is the given URL's host (or a parent domain) in the blocklist set?
 * Matches the exact host and any parent domain (so `evil.com` blocks
 * `sub.evil.com`).
 */
export function isHostBlocked(rawUrl: string, set: Set<string>): boolean {
  let host: string
  try {
    host = normalizeHost(new URL(rawUrl).hostname)
  } catch {
    return false // malformed URLs are rejected earlier by validation, not here
  }
  if (set.has(host)) return true
  // Walk parent domains.
  const parts = host.split('.')
  for (let i = 1; i < parts.length - 1; i++) {
    if (set.has(parts.slice(i).join('.'))) return true
  }
  return false
}

/** Convenience wrapper using the loaded file-backed set. */
export function isUrlBlocked(rawUrl: string): boolean {
  return isHostBlocked(rawUrl, loadSet())
}

/** Test/seed hook to inject a set (resets the file-backed cache). */
export function __setBlocklistForTest(set: Set<string> | null): void {
  cachedSet = set
}

File: backend/src/services/blocklist.rs
//! Inbound (create-time) phishing/malware blocklist — ported from
//! `src/lib/blocklist.ts`. Offline newline-delimited host file loaded once into
//! an in-memory set. INBOUND boundary, separate from the outbound SSRF guard.

use std::collections::HashSet;
use std::sync::OnceLock;

/// Normalize a host: lowercase, strip a leading `www.` and a trailing dot.
fn normalize_host(host: &str) -> String {
    let lower = host.to_lowercase();
    let no_www = lower.strip_prefix("www.").unwrap_or(&lower);
    no_www.strip_suffix('.').unwrap_or(no_www).to_string()
}

/// Parse blocklist file text into a normalized host set. Pure.
pub fn parse_blocklist(text: &str) -> HashSet<String> {
    let mut set = HashSet::new();
    for line in text.split('\n') {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        set.insert(normalize_host(trimmed));
    }
    set
}

/// Pure: is the given URL's host (or a parent domain) in the blocklist set?
/// Matches the exact host and any parent domain (so `evil.com` blocks
/// `sub.evil.com`).
pub fn is_host_blocked(url: &url::Url, set: &HashSet<String>) -> bool {
    let host = match url.host_str() {
        Some(h) => normalize_host(h),
        None => return false,
    };
    if set.contains(&host) {
        return true;
    }
    let parts: Vec<&str> = host.split('.').collect();
    // Walk parent domains: i in [1, parts.len()-1)
    if parts.len() >= 2 {
        for i in 1..parts.len() - 1 {
            let candidate = parts[i..].join(".");
            if set.contains(&candidate) {
                return true;
            }
        }
    }
    false
}

static BLOCKLIST: OnceLock<HashSet<String>> = OnceLock::new();

/// Load the file-backed set once from `data/blocklist.txt` (relative to CWD),
/// falling back to an empty set when absent or unreadable.
fn load_set() -> &'static HashSet<String> {
    BLOCKLIST.get_or_init(|| {
        let path = std::path::Path::new("data").join("blocklist.txt");
        match std::fs::read_to_string(&path) {
            Ok(text) => parse_blocklist(&text),
            Err(_) => HashSet::new(),
        }
    })
}

/// Is this URL blocked according to the loaded file-backed set?
pub fn is_blocked(url: &url::Url) -> bool {
    is_host_blocked(url, load_set())
}

#[cfg(test)]
mod tests {
    use super::*;
    use url::Url;

    fn set(entries: &[&str]) -> HashSet<String> {
        entries.iter().map(|s| s.to_string()).collect()
    }

    #[test]
    fn parse_strips_comments_and_blanks_and_www() {
        let text = "# comment\n\nwww.Evil.com\nbad.org.\n  spaced.net  \n";
        let s = parse_blocklist(text);
        assert!(s.contains("evil.com"));
        assert!(s.contains("bad.org"));
        assert!(s.contains("spaced.net"));
        assert_eq!(s.len(), 3);
    }

    #[test]
    fn matches_exact_and_subdomains() {
        let s = set(&["evil.com"]);
        assert!(is_host_blocked(&Url::parse("http://evil.com/x").unwrap(), &s));
        assert!(is_host_blocked(&Url::parse("https://sub.evil.com/x").unwrap(), &s));
        assert!(is_host_blocked(&Url::parse("https://www.evil.com/").unwrap(), &s));
        assert!(is_host_blocked(&Url::parse("https://a.b.evil.com/").unwrap(), &s));
    }

    #[test]
    fn does_not_match_unrelated() {
        let s = set(&["evil.com"]);
        assert!(!is_host_blocked(&Url::parse("https://notevil.com/").unwrap(), &s));
        assert!(!is_host_blocked(&Url::parse("https://example.com/").unwrap(), &s));
        // parent-walk must not match the TLD alone
        let s2 = set(&["com"]);
        assert!(!is_host_blocked(&Url::parse("https://evil.com/").unwrap(), &s2));
    }
}
